### ISSUE TITLE
ghcide-tests' `addDependentFile` test

### ISSUE THREAD

**Issue by soulomoon** [haskell/haskell-language-server#4194]
What is happenning
* GetModificationTime dep-file.txt cache value in shake extra is cleanup
* GetModificationTime dep-file.txt is dirty
* shake restart
* GetModificationTime dep-file.txt computed
* GetModificationTime dep-file.txt cache value is added
* but the new value is not flush the result to hls-graph (Hence its version step is not updated)
* shake restart
* computing GetModificationTime dep-file.txt
* GetModificationTime dep-file.txt cache value is read (The source of the error) and returned
* it's reverse dep compute, since its step is not updated in the last run,  the value is considered as the old result, so it's reverse dep is not recomputed.

============================
After file change and index
The correct case
```
<-- {
    "jsonrpc": "2.0",
    "method": "$/progress",
    "params": {
        "token": "22",
        "value": {
            "kind": "end",
            "message": "Finished indexing 1 files"
        }
    }
}
<-- {
    "jsonrpc": "2.0",
    "method": "textDocument/publishDiagnostics",
    "params": {
        "diagnostics": [],
        "uri": "file:///private/var/folders/36/_743psv11gv2wrj9dclrpd500000gn/T/hls-test-root/extra-dir-52959961234/Foo.hs",
        "version": 0
    }
```

but we sometimes we have, never get the diganostic for file that have changed 
```
<-- {
    "jsonrpc": "2.0",
    "method": "$/progress",
    "params": {
        "token": "23",
        "value": {
            "kind": "end",
            "message": "Finished indexing 1 files"
        }
    }
}
```

**Comment by soulomoon** [comment]
failed
```
"defineEarlyCutoff' start ReportImportCycles Foo.hs"
"defineEarlyCutoff' start ReportImportCycles Baz.hs"
"New diagnostics for ReportImportCycles Baz.hs: []"
"defineEarlyCutoff' remove dirty ReportImportCycles Baz.hs"
"New diagnostics for ReportImportCycles Foo.hs: [Diagnostic {_range = Range {_start = Position {_line = 4, _character = 11}, _end = Position {_line = 7, _character = 52}}, _severity = Just DiagnosticSeverity_Error, _code = Just (InR \"-Wdeferred-type-errors\"), _codeDescription = Nothing, _source = Just \"typecheck\", _message = \"\\8226 Couldn't match type \\8216[Char]\\8217 with \\8216Int\\8217\\n  Expected: Int\\n    Actual: String\\n\\8226 In the second argument of \\8216(+)\\8217, namely \\8216(\\\"A\\\")\\8217\\n  In the expression: 1 + (\\\"A\\\")\\n  In an equation for \\8216foo\\8217: foo = 1 + (\\\"A\\\")\", _tags = Just [], _relatedInformation = Just [DiagnosticRelatedInformation {_location = Location {_uri = Uri {getUri = \"file:///private/var/folders/36/_743psv11gv2wrj9dclrpd500000gn/T/hls-test-root/extra-dir-58672764745/Foo.hs\"}, _range = Range {_start = Position {_line = 4, _character = 11}, _end = Position {_line = 7, _character = 52}}}, _message = \"TypeCheck\"}], _data_ = Nothing}]"
"defineEarlyCutoff' remove dirty ReportImportCycles Foo.hs"
"defineEarlyCutoff' start GhcSessionDeps Foo.hs"
"defineEarlyCutoff' remove dirty GhcSessionDeps Foo.hs"
"defineEarlyCutoff' start GetModificationTime dep-file.txt"
"New diagnostics for GetModificationTime dep-file.txt: []"
"defineEarlyCutoff' start GetModIface Foo.hs"
"defineEarlyCutoff' remove dirty GetModificationTime dep-file.txt"
"defineEarlyCutoff' start TypeCheck Foo.hs"
"New diagnostics for TypeCheck Foo.hs: [Diagnostic {_range = Range {_start = Position {_line = 4, _character = 11}, _end = Position {_line = 7, _character = 52}}, _severity = Just DiagnosticSeverity_Error, _code = Just (InR \"-Wdeferred-type-errors\"), _codeDescription = Nothing, _source = Just \"typecheck\", _message = \"\\8226 Couldn't match type \\8216[Char]\\8217 with \\8216Int\\8217\\n  Expected: Int\\n    Actual: String\\n\\8226 In the second argument of \\8216(+)\\8217, namely \\8216(\\\"A\\\")\\8217\\n  In the expression: 1 + (\\\"A\\\")\\n  In an equation for \\8216foo\\8217: foo = 1 + (\\\"A\\\")\", _tags = Just [], _relatedInformation = Just [DiagnosticRelatedInformation {_location = Location {_uri = Uri {getUri = \"file:///private/var/folders/36/_743psv11gv2wrj9dclrpd500000gn/T/hls-test-root/extra-dir-58672764745/Foo.hs\"}, _range = Range {_start = Position {_line = 4, _character = 11}, _end = Position {_line = 7, _character = 52}}}, _message = \"TypeCheck\"}], _data_ = Nothing}]"
"defineEarlyCutoff' remove dirty TypeCheck Foo.hs"
```
succ 
```
"defineEarlyCutoff' start ReportImportCycles Foo.hs"
"defineEarlyCutoff' start ReportImportCycles Baz.hs"
"New diagnostics for ReportImportCycles Baz.hs: []"
"defineEarlyCutoff' remove dirty ReportImportCycles Baz.hs"
"New diagnostics for ReportImportCycles Foo.hs: [Diagnostic {_range = Range {_start = Position {_line = 4, _character = 11}, _end = Position {_line = 7, _character = 52}}, _severity = Just DiagnosticSeverity_Error, _code = Just (InR \"-Wdeferred-type-errors\"), _codeDescription = Nothing, _source = Just \"typecheck\", _message = \"\\8226 Couldn't match type \\8216[Char]\\8217 with \\8216Int\\8217\\n  Expected: Int\\n    Actual: String\\n\\8226 In the second argument of \\8216(+)\\8217, namely \\8216(\\\"A\\\")\\8217\\n  In the expression: 1 + (\\\"A\\\")\\n  In an equation for \\8216foo\\8217: foo = 1 + (\\\"A\\\")\", _tags = Just [], _relatedInformation = Just [DiagnosticRelatedInformation {_location = Location {_uri = Uri {getUri = \"file:///private/var/folders/36/_743psv11gv2wrj9dclrpd500000gn/T/hls-test-root/extra-dir-91307068645/Foo.hs\"}, _range = Range {_start = Position {_line = 4, _character = 11}, _end = Position {_line = 7, _character = 52}}}, _message = \"TypeCheck\"}], _data_ = Nothing}]"
"defineEarlyCutoff' remove dirty ReportImportCycles Foo.hs"
"defineEarlyCutoff' start GhcSessionDeps Foo.hs"
"defineEarlyCutoff' remove dirty GhcSessionDeps Foo.hs"
"defineEarlyCutoff' start GetModificationTime dep-file.txt"
"defineEarlyCutoff' start GetModIface Foo.hs"
"New diagnostics for GetModificationTime dep-file.txt: []"
"defineEarlyCutoff' remove dirty GetModificationTime dep-file.txt"
"defineEarlyCutoff' start TypeCheck Foo.hs"
"New diagnostics for TypeCheck Foo.hs: []"
"defineEarlyCutoff' remove dirty TypeCheck Foo.hs"
```

**Comment by soulomoon** [comment]
Odd is that even 

```
```

**Comment by soulomoon** [comment]
`GetModificationTime` in hls-graph is not updated correctly

**Comment by soulomoon** [comment]
We should be deleting the `GetModificationTime` for watched file, but some how we did not always finish
https://github.com/soulomoon/haskell-language-server/blob/9dc0281413eb5cc92aa4d5d912691646ca142fa5/ghcide/src/Development/IDE/Core/FileStore.hs#L167

**Comment by soulomoon** [comment]
`defineEarlyCutoff'` remove dirty GetModificationTime "dep-file.txt" but it's value never reach to actually update hls-graph

**Comment by soulomoon** [comment]
The solution would be the removal of dirty keys should be running after hls-graph update the key value

**Comment by soulomoon** [comment]
fixed by #4190
